### ISSUE TITLE
001 NCSM - Unable to change cycle on Atlanta NCSM account

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#892]
In GitLab by @kdjstudios on May 9, 2018, 09:38

**Submitted by:** "Jesus Corchado" <jesus.corchado@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2018-05-09-93738/conversation

**Server:** Internal

**Client/Site:** NCSM - 53

**Account:** N0020

**Issue:** 

I am having trouble with one account in Atlanta.  I tried changing the cycle several times to the NCSM 1st cycle accounts but every time I click save, it does not make the change as stays as it was before.  Can you please help me out on this one?  It’s for account 053-N0020.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 9, 2018, 09:40

@smasih @tim.traylor @pchaudhary @thomas.gentry 

I wrote:

> Thank you for bringing this to our attention. I am also unable to change the Billing cycle for this account. We have notified the support team and will follow up accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 10, 2018, 04:53

One of the account activity `053 NCSM Usage` was having an empty field which was stopping the account from being saved.

Another validation issue. We could have avoided this ticket if the user was presented with proper message. 

@kdjstudios @tim.traylor we should pipeline the validation changes.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 10, 2018, 12:04

@smasih @tim.traylor @pchaudhary @rchauhan 

Which field was empty? How did you fix it? Please specify these details. This lack of detail is critical and can end up causing delays; as well can cause this issue to continue if we are unable to inform operations of the corrections.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 10, 2018, 12:06

I wrote:

> The support team was able to look into this account and resolve it. Please review and confirm everything is correct. Thank you.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on May 11, 2018, 06:10

@smasih Sumeet, please follow up and let us know which field was missing and what validation needs to be added.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 23, 2018, 15:08

@smasih @pchaudhary @rchauhan 

Please provide the requested information so we may proceed with marking this ticket as closed.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 24, 2018, 01:08

@kdjstudios 

There hasn't been any progress on this ticket, we will try to wrap this up in next cycle.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 29, 2018, 05:02

Hi [Tim] ( @tim.traylor ) & [Kyle] ( @kdjstudios ),  

> Sumeet, please follow up and let us know which field was missing and what validation needs to be added.

* The *activity_id name* was missing from one of the activities of the account due to which that all happened, so to prevent the same in future, we can add a validation for all of the fields (except description) on the activity & recurring fees at the time of being saved.

We have looked into this & seems like the simple model validation will not work here as the activity is embedded in the account, so updating the estimate accordingly.

`estimate 2-3hr`

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 4, 2018, 05:57

@kdjstudios Can you pls check w/Ops and confirm that the stats multiplier would NEVER be a negative value?  Thx

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 4, 2018, 09:42

@tim.traylor 

I wrote an email to Cori to confirm if the stats multiplier value will ever be a negative value.

HD: http://www.servicedesk.answernet.com/profiles/ticket/2018-06-04-59571/conversation

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 4, 2018, 23:11

assigned to @pchaudhary

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 5, 2018, 05:05

Hi [Kyle] ( @kdjstudios ),   

We have almost completed the development of this. The CSS part is remaining. We will wait for the response from Cori for further move.

**Branch** - 908_dup_account_no
**Server** - Merge2

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 6, 2018, 01:49

marking it paused till then

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 6, 2018, 10:12

@smasih @pchaudhary @tim.traylor 

According to @tim.traylor during the Tuesday meeting we are to request this operational information directly from Gary. Marking for discussion. Please be sure to bring up on Thursday.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 7, 2018, 05:23

**in nutshell we need to discuss for account validation if "Stats Multiplier" should always be positive.**

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 8, 2018, 24:18

@kdjstudios 

Gary, confirmed that SM should be always positive.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 8, 2018, 07:58

Hi [Kyle] (@kdjstudios ),   

We have completed the development on this. We have successfully applied the validation checks as follows: 

##### Recurring Fees  
*  **Name** cannot be saved as blank.  
*  **Quantity** cannot be saved smaller than or equal to zero.  

##### Activities  
*  **Name** cannot be saved as blank.  
*  **Allowance** cannot be saved as blank.  
*  **SM** cannot be smaller than or equal to zero.   

##### Rebillables  
*  **Vendor** & **Vendor Service** cannot be saved as blank.  
*  **Quantity** cannot be saved smaller than or equal to zero.  


Please look into the Merge2 server for the testing purpose & mark as ready if all goes well.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 8, 2018, 10:14

@pchaudhary @smasih 

Thank you for the updates. Please let us know what testing you have performed and the results. Please let us know what you would like us to test and how. Thank you.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 8, 2018, 15:32

Marking for testing.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 9, 2018, 02:05

@kdjstudios 

> Thank you for the updates. Please let us know what testing you have performed and the results. Please let us know what you would like us to test and how. Thank you.

We always do a developer level testing before pushing it forward for your testing and approval.

Here, we need to check validation on **"account edit page"** so please be random, try leaving fields blank or filling incorrect values and see how the system reacts and guides you to fix the issue.

For eg: 

I am trying to save activities on account with a blank value.

![image](/uploads/022b67d4133e135e178a322016d9dff7/image.png)

System, will ask you to fill the field.

![image](/uploads/79fea2b730b7706bc9625718a6722ab1/image.png)

If we leave allowance blank

![image](/uploads/e77db3b9a4cbeb4ab361169f814d0621/image.png)

Tried to save blank account number.

![image](/uploads/0d140a09ca279cdb8bddc7d0a36c31ba/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 9, 2018, 16:54

@smasih @pchaudhary @tim.traylor 

Thank you for the update and clarification. Will test on Monday.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 11, 2018, 11:52

@smasih @pchaudhary @tim.traylor 

Was able to update

## Recurring Fees

- **Name** cannot be saved as blank. 

> Success

- **Quantity** cannot be saved smaller than or equal to zero.

> FAIL 
> - We will want the note next to the field to be red in color like the other notes

## Activities

- **Name** cannot be saved as blank.

> Success

- **Allowance** cannot be saved as blank.

> Fail:
> - We will want the note next to the field to be red in color like the other notes*
> - Seems to only report one error when there are multiple. Needs to display all errors.
> - When Excess Rate and/or Description is left blank just says activity is invalid. Needs to highlight field(s) red and show note like for quantity and allowance. 
> - Also now when successfully updated with information; it seems still shows as invalid, but I can not determine what is invlaid??
> ![image](/uploads/b2a3ca2053f9eecc1a3eae77682126ca/image.png)
> ![image](/uploads/d2604b071272a12ab475a4b261e76b00/image.png)
> - Tried to 'remove/deactivate' the fee and received this error and the system got stuck on a loading screen then. 
> ![image](/uploads/18392214f3132f2f71f2e8fa652d6ab8/image.png)
>

- **SM** cannot be smaller than or equal to zero.

> This should not be visible to anyone other then tech admins correct?

## Rebillables

- **Vendor** & **Vendor Service** cannot be saved as blank.

> FAIL
> Just clicked "+Add vendor Service"; left everything blank and was able to successfully create vendor service
> ![image](/uploads/54accc09ec8ba77526e254539412bbb2/image.png)

- **Quantity** cannot be saved smaller than or equal to zero.

> FAIL
> Was able to update the quanity to negative 5 without error.
>![image](/uploads/16420dfc7ee4b708ce2a2354569e2a84/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 12, 2018, 07:36

Hi [ Kyle] ( @kdjstudios ),   

Thanks for your testing.  


#### Recurring Fees
------

*  All goes well for us.  

#### Activities
-------
*  All goes well for us.  

> This should not be visible to anyone other than tech admins correct?

* Correct. It's already in the code.

####  Rebillables
-------

> FAIL, Just clicked "+Add vendor Service"; left everything blank and was able to successfully create vendor service

* Earlier, we did not validate the field for that page in which you were testing (Main Vendor Page) but thanks to notifying for that, we have validated it now, however, we were talking about the account vendor services in the rebillables section of the account edit page.


**Note-** We understand your concern that we should follow the uniform validation rule but as per the previous discussion on the SAB validation rules we should follow the server validation instead of client validation & there are some limitations in that due to which we can't do that validation untill the **SAVE** button click.  

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 12, 2018, 10:33

During discussion it was mentioned that we need uniform validation across all fields and display all errors at once.
- Please highlight input box/field in red color
- Please display "Invalid value" in red color text notice; in place of "can't be blank"

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 13, 2018, 08:23

Hi [Kyle] ( @kdjstudios ),   

As suggested by [Tim] (@tim.traylor ) on yesterday's call to validate the fields by using any client validation type.

We have applied the same to the accounts page. There are some limitations on the form due to which we are not able to validate the quantity fields to be greater than 0 but that is checking by the server side validation.
Please test on Merge2 Server.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 13, 2018, 08:45

@pchaudhary @tim.traylor @smasih 

Thank you, this client side validation seems 20 times quicker then it was before. However, we are still having the same issues:

- Only showing one error, not all errors.
- Not displaying in RED Color.
- Why does it say "This field is required." and not "Invalid value" like requested?

![image](/uploads/60a64b5aa2edc1e842b24658670fd614/image.png)


*Needs to display like:*
![image](/uploads/99329b1d98e2b3be4eaaf195089add17/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 14, 2018, 01:45

Hi [Kyle] ( @kdjstudios ),  

Thanks for your testing.

> Only showing one error, not all errors.

* I'm able to see all errors at once.  
 The case you were testing only had one validation because the *excess rate* has a default value of `0.00` & *description* does not have any validation.  


> Not displaying in RED Color.

* Completed now.

> Why does it say "This field is required." and not "Invalid value" as requested?

* Completed.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 14, 2018, 08:02

Hi [Kyle] ( @kdjstudios ),  

All of the client-side validation on the account, customer, vendor, vendor-services have been done successfully. Please test on Merge2 Server.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 20, 2018, 05:42

Hi [Kyle] ( @kdjstudios ),   

Can you please provide an update on this?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 20, 2018, 10:49

@pchaudhary @tim.traylor 

I have tested and overall there are lots of good improvements and a majority of successful functionality. But below are all of the issues I could find.

# Account:

- Should the resource ID be required? I though this is needed for the usage tracking, billing, etc?
- recurring fees:
 - Only provided an error for the drop down. 
 - I would believe we need to have at least the base rate and the billing period be required too? 
 - Maybe the description too? or should that default to the fee name?
- Account activities:
 - Still not showing error in red
 - Only shows one error at a time
 - I would believe we need to have at least the excess rate be require too?
 - Maybe the description too? or should that default to the activity name?
- CC on file:
 - No errors for all of the required fields; when left blank. 
 - No errors on the "get details" link too; when leaving fields blank. 

# Customer:

- Tried to enter information into the address (after receiving the invalid error) and then the "Copy" link does not work.
- was able to save the customer successfully, then clicked edit and tried to save again; received the "something went wrong error"
- Tried to send portal invite with the email being blank; received the "something went wrong error"

# Vendor:
- Name and account number should be required fields.
- Errors were not in RED.
- Error Fields were not highlighted red with red error next to it. 

# Vendor Services:
- UOM needs to be required.
- Errors were not in RED. 
- Error Fields were not highlighted red with red error next to it. 
- Tried to save with only putting in a name; received the "something went wrong error"


# Site:
- Removed the site Number from the demo site, received an error,; changed its value and then saved. The site no longer shows in drop down, search, site page, etc.
 - Changed the site number back to the original value and now it shows back up. Should we not allow for changing the site number through the GUI?
- Possible required fields: Site name, Site number, billing cycle, day of month, service start date,last billing cycle date, terminated account prefix, 
- Error displayed "trying to load additional fees"; but there are none on a new site?
- Once it appeared to save successfully;
 - Added the FTP page, the LDAP, and user permission, but I can not find the site "VTest" in the drop down, search, site page, etc.
- Could not continue processing testing at the site level due to it disappearing. lol

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 20, 2018, 10:51

Also most errors still are showing "can't be blank" When they should be showing "Invalid Value" as pointed out in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/892#note_22103

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 21, 2018, 06:01

Hi [Kyle] ( @kdjstudios ),  

Thanks for your testing results. We have made some improvements based on your results. Please check.

# Account:  

> Should the resource ID be required? I thought this is needed for the usage tracking, billing, etc?
  * Yes, thanks for reminding. Now, we have added the 2 types of validation on for resource ID. One is to be **presence** & another is to be **unique** under the scope of its Site.

> ##### Recurring fees:  
  * Only provided an error for the drop down.   
    * It's running perfectly in our system. Please provide any further details. 
  * I would believe we need to have at least the base rate and the billing period be required too?  
    * We understand your concern but the system is designed in such a way that if the billing period of any recurring fees is **blank** then it'll include on **each & every invoice of the account**. So, we suggest to not change that functionality right now. And the base rate has **0** as the default value so we are not validating it on the client level. But if need, then we can add a server-side validation to check the input value must be greater than **0**.   
Jotting down for discussion.
  * Maybe the description too? or should that default to the fees name?  
    * Jotting down for discussion.  

> ##### Account Activities:
  * Still not showing error in red. 
    * It's running perfectly in our system. Please provide any further details.  
  * Only shows one error at a time. 
    * It's running perfectly in our system. Please provide any further details.   ![image](/uploads/18b00cb0048290b409b0df7b47f1324d/image.png)
  * I would believe we need to have at least the excess rate requires too?  
    * The excess rate has **0** as the default value so we are not validating it on the client level. But if need, then we can add a server-side validation to check the input value must be greater than **0**.  
Jotting down for discussion.
  * Maybe the description too? or should that default to the activity name?  
    * Jotting down for discussion.  

> ##### CC On File:
  * No errors for all of the required fields, when left blank.  
  * No errors on the "get details" link too; when leaving fields blank.
    * I think we shouldn't add a validation on CCOF section, however, we can discuss further with Tim.  

# Customer:

> Tried to enter information into the address (after receiving the invalid error) and then the "Copy" link does not work.  

* Please provide the specific steps you performed. We were not able to get it.  

> Was able to save the customer successfully, then clicked edit and tried to save again; received the "something went wrong error"  

* Please provide any further details. We are unable to reproduce the error.
  
> Tried to send a portal invite with the email being blank; received the "something went wrong error"

* Thanks to spot. Resolved now.

# Vendor: 

> Name and account number should be required fields.

* **Vendor Name** are already validated and now we have also added a validation for **Vendor Account Number**.

> Errors were not in RED.  
> Error Fields were not highlighted red with red error next to it.

* They are red in our systems. Please provide any further details. ![image](/uploads/07e5e2b2e595f5e1fcf6c2b6f0e9bdbb/image.png)

# Vendor Services:

> UOM needs to be required.

* Thanks for reminding. Successfully added the validation for UOM. ![image](/uploads/710fa6e3fd77d7534181c3fbc7ee0572/image.png)

> Tried to save with only putting in a name; received the "something went wrong error"

* Thanks to pointing it out. Resolved.

# Site

* We have not done any validation on the Site level yet.

> Also most errors still are showing "can't be blank" When they should be showing "Invalid Value" as pointed out in.

* As per our knowledge, only the server-side validation stated "can't be blank", if any client side-validation is also showing the same message then please let us know the place.


# Points to be discussed with Tim:
##### Recurring Fees:
1.  Billing Period and Base Rate validation
1.  Description field validation.

##### Account Activities:
1. Excess Rate validation.
1. Description field validation.

##### CCOF
1. CCOF Validation.


Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 21, 2018, 10:25

@smasih @pchaudhary @tim.traylor 

From discussion today:

# Recurring fees:
- Billing Period: Please allow a 'NA' option by default.
- Base rate: Needs to be required. Needs to be a numerical value.
- Quantity: Needs to be required. Needs to be a numerical value
- Description: Needs to be required. 

# Activities:
- Adding new the trash can, needs boundaries. 
- Needs to fix header alignment with the field positions. #941 
- Description: Needs to be required. Also move next to the name column to better align with other options. (Name/Description/Excess rate/allowance/taxable/SM/Active)
- Excess rate: Needs to be required. Needs to be a numerical value.
- Allowance: Needs to be required. Needs to be a numerical value.

# CCOF
- Please ignore until we have updated the CCOF process.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 21, 2018, 10:37

> Tried to enter information into the address (after receiving the invalid error) and then the "Copy" link does not work.

- Please provide the specific steps you performed. We were not able to get it.
 - Thank you for updating this it is now working on MERGE2. It was previously mentioned in #941 with a screenshot of the issue too. Since these changes are not live yet on production they are still experiencing the issue on production. 

> Was able to save the customer successfully, then clicked edit and tried to save again; received the "something went wrong error"

- Please provide any further details. We are unable to reproduce the error.
 - Click "New customer"
 - Click "Create Customer" (Do not enter any information in)
 - You will be presented with errors. Correct the errors. Click "Update Customer"
 - Should now save successfully. Click "Edit Customer"
 - You will now be presented with "Something went wrong"
 - Here is the link: https://merge2.sabilling.com/customers/5b2bd2b70be3e35e75000012
 - You now have to click the back button in the web browser and then refresh page, then click "Edit Customer" to get it to work. 

> Also most errors still are showing "can't be blank" When they should be showing "Invalid Value" as pointed out in.

- As per our knowledge, only the server-side validation stated "can't be blank", if any client side-validation is also showing the same message then please let us know the place.
 - I see the updated "Invalid Values" and they are finally in RED!! Awesome Way to go!

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 22, 2018, 06:22

Hi [Kyle] (@kdjstudios ),   

Thanks for your updates. We have done some changes according to the requirements. Please check.  

# Recurring Fees: 

> Billing Period: Please allow an 'NA' option by default.

* Completed.

> Base rate: Needs to be required. Needs to be a numerical value.  

* Added a Validation.

> Quantity: Needs to be required. Needs to be a numerical value

* Added a Validation.

> Description: Needs to be required.

* Added a Validation.

# Activities:

> Adding new the trash can, needs boundaries.

* Completed.

> Needs to fix header alignment with the field positions. #941

* Completed.

> Description: Needs to be required. Also move next to the name column to better align with other options. (Name/Description/Excess rate/allowance/taxable/SM/Active)

* Completed.

> Excess rate: Needs to be required. Needs to be a numerical value.

* Completed.

> Allowance: Needs to be required. Needs to be a numerical value.

* Completed.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 22, 2018, 06:54

Hi [Kyle] ( @kdjstudios ),   


> Please provide any further details. We are unable to reproduce the error.
>
> - Click "New customer"
>
> - Click "Create Customer" (Do not enter any information in)
>
> - You will be presented with errors. Correct the errors. Click "Update Customer"
>
> - Should now save successfully. Click "Edit Customer"
>
> - You will now be presented with "Something went wrong"
>
> - Here is the link: https://merge2.sabilling.com/customers/5b2bd2b70be3e35e75000012
>
> - You now have to click the back button in the web browser and then refresh the page, then click "Edit Customer" to get it to work.

* Thanks for the steps. It really helps us to reach the cause of the issue. It is resolved now. Please check.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 22, 2018, 10:20

@pchaudhary 

Thank you for the update: Looks like we are down to one issue; the description field for both Recurring fees and activities are still not showing as required and presenting an error. 

![image](/uploads/b41506e0f5c761fbf1f48b5211d59ae6/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 23, 2018, 01:24

@kdjstudios 

There are currently two level of validations:  
 * Client-side (on page validation)  
 * Server-side   

description column on activities and recurring fees is validated on the server because we have to let it pass if the **"blank description column"** belongs to an inactive fee or activity (deactivated at site or account level).   

Server-side validations comes in action once client's side validation is complete.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 25, 2018, 08:55

Hi [Kyle] (@kdjstudios ),   

Please do not test it today because we've found an error. We will resolve it by tomorrow.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 28, 2018, 07:54

Hi [Kyle] (@kdjstudios ),  

Please test it today on the Merge2 server & forward it for the release.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 28, 2018, 14:26

@pchaudhary 

Thanks everything appears to be showing the errors and not allowing for saving of invalid information. the only thing I see is that we are still showing "can't be blank" instead of the "Invalid Value" in some of the errors. 

![image](/uploads/efca322f7a08a180e2e4c87af584a7d4/image.png)


@tim.traylor 

Also at this same point in the screen shot above, I can not delete those activities. It will only allow me to deactivate. So my question here, it would appear that maybe we are still allowing for saving of invalid information? As we need to be able to delete items that are not correct; not have to correct them and then deactivate them. Seems like we will end up with lots of invalid or inactive items this way.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 29, 2018, 06:26

Hi [Kyle] ( @kdjstudios ),  

Thanks for your testing update.  

> The only thing I see is that we are still showing "can't be blank" instead of the "Invalid Value" in some of the errors.  

* The "can't be blank" is the default message comes up only after the server-side validation came into the action. Albeit, we can change the default message if you want.  


> Also at this same point in the screenshot above, I can not delete those activities. It will only allow me to deactivate. So my question here, it would appear that maybe we are still allowing for saving of invalid information? As we need to be able to delete items that are not correct; not have to correct them and then deactivate them. Seems like we will end up with lots of invalid or inactive items this way.

* We will discuss with Tim about this & will post an update.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 6, 2018, 09:23

@pchaudhary 

Mentioned that this was already discussed with @tim.traylor today and there is a plan of action to insure until an item is valid it can not be saved.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 12, 2018, 05:37

There is a bug reported by [Kyle] (@kdjstudios ) while after the site change.  
> I edited account # then selected the site drop down and choose a different site. I then clicked okay and choose to save. but was presented with the "Something went wrong error". (https://scc.sabilling.com/accounts/5ac8b91497a7379f58000005) I clicked the back button and then clicked to edit the account. It appears to have saved, because all fees, activities, etc were removed.  

We have to look into this and make it correct along with this issue.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 12, 2018, 08:56

@pchaudhary 

Just to summarize; we have two issues for this ticket that are needing to be addressed. Please let me know if I missed anything.

1. Fees and activities will use a client side check on all fields to insure that only valid information is saved, and that until that happens the user can still delete the fee.
2. Error when trying to save site after changing site from drop down.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 13, 2018, 09:24

Hi [Kyle] ( @kdjstudios ),  

There are total 3 issues for this ticket that are needing to be addressed.  
  1. Fees and activities will use a client-side check on all fields to ensure that only valid information is saved and that until that happens the user can still delete the fee.  
    * This is remaining. We have to work on this.  
  2. Error when trying to save site after changing the site from drop down.  
    * This is remaining.  
  3. Need a confirmation box rather than alert while switching the site.  
    * It's completed.  

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 16, 2018, 07:37

Hi [Kyle] ( @kdjstudios ),  

We have fixed the 2^nd point as well.  

* Error when trying to save site after changing the site from drop down.

Fixed

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 16, 2018, 09:52

@pchaudhary 

Thank you for the updates. Please let us know when to test that the Fees and activities use a client-side check on all fields.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 19, 2018, 04:35

Hi [Kyle] (@kdjstudios ),   

We have completed the development on this ticket & you can go for the testing for, if approved then we can release this on upcoming Saturday.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 19, 2018, 11:29

@pchaudhary 

Testing steps:

1. Selected account "Kyle 3" 
2. Clicked edit account.
3. Choose the demo site from the drop down
4. Clicked "Ok" on the pop up
5. Clicked "Save" on the account
6. Received error: https://merge2.sabilling.com/accounts/5ac8cd1a97a7378e6f000012

Testing steps(different account)

1. Selected account "Test Dup" 
2. Clicked edit account.
3. Selected button to add Activity and add recurring fee. Left all prompts blank
4. Clicked "Save"
5. Appears that the "Description" is still not being checked on the client side, as no error is shown. 

![image](/uploads/6fcd662415bae71f82510b9721592a61/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 20, 2018, 01:46

Hi [Kyle] ( @kdjstudios ),   

We have tested this on our end & all goes well.  

* The Site Change functionality is running well, but the account you tested is the Test account so it does have some issues with the activities due to which it was throwing an error, please check now & if the issue persists then go with some other account.

> Appears that the "Description" is still not being checked on the client side, as no error is shown.

* I understand your concern on the above points & here are the few things we would like to inform you for the clear understanding of that:
  1. There are some activities/fees on the production servers which are currently deactivated and also don't have any description on it. So in this situation, if any user wants to change any field of the account & try to save them then it will throw a warning message on those activities/fees description field but the user can't be able to edit those because they are non-editable (because of the deactivation). For overcome this issue we have validated the description column on the Server Side, we are validating only those activites/fees descriptions which are currently activated at the time of account save.  
Eg- For the clear picture you can go on this link => [Speed Ez's account](https://merge2.sabilling.com/accounts/56d4bb8997a737b7d2000104/edit?sid=55d4ba2de13823611f000003) & can see that the deactivated activities/fees do not have descriptions & the user can't edit to them until the activation of those.

Hope this helps.  
Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 20, 2018, 11:32

@pchaudhary 

Please have the description check on the client side. If you recall a few weeks ago during the call this was discussed and that was the solution we requested. I believe http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/892#note_23160 is the comment that was posted after that discussion. 

Also now please provide an list of the activities and fees that do not have a description as this should not be allowed and we will need to correct those too now. Thank you for the example, please keep in mind that account was terminated/setup before we had these requirements in place, so we will need that list to them reach out to all effected clients sites and have those corrected too. I will create a new ticket for this task (#992) please post all updates regarding those to that ticket.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 21, 2018, 24:37

@kdjstudios 

sure, but let's discuss it with Tims as well.

marking for discussion.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 23, 2018, 01:25

Hi [Kyle] (@kdjstudios  ),  

> Please have the description check on the client side. If you recall a few weeks ago during the call this was discussed and that was the solution we requested. I believe http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/892#note_23160 is the comment that was posted after that discussion.

* Yes, we remember that but we inkling that we had mentioned about the description which is going to be validated by the backend due to the noticed reason, that's why we mentioned this thing in our above comments { [comment 1] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/892#note_20771), [comment 2] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/892#note_22563)} every time.  



> Also now please provide a list of the activities and fees that do not have a description as this should not be allowed and we will need to correct those too now.

* Here it is.   
  [answernet_accounts_having_blank_description.csv](/uploads/07813ff011b86d9a1a1ba47c0af16d1d/answernet_accounts_having_blank_description.csv)  
  [hosted_accounts_having_blank_description.csv](/uploads/d07159efb900c10657a2d1bf39e713c3/hosted_accounts_having_blank_description.csv)

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 23, 2018, 10:25

@pchaudhary @smasih 

Please bring this back up in discussion tomorrow. As I recall the **latest** update http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/892#note_23160 was that we need to have client side check for the description.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 24, 2018, 09:51

@pchaudhary @smasih @tim.traylor 

During the call today, Tim mentioned that we should be able to do a client side validation on the description check.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 26, 2018, 02:46

@kdjstudios @tim.traylor 

Added the description validation on the client side as discussed on the tech call. (there is a validation on server side as well)

Please Test on **merge2 server**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 26, 2018, 08:21

@smasih 

Thank you Looks great! marking for ready

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 2, 2018, 23:58

primary of #992

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 3, 2018, 09:04

@smasih 

Attaching the spreadsheet form the Ruby testing to insure all items of needed correction are corrected with this release. Please insure all items that are marked with 892 tick number are working correctly before you release. 

[Ruby_Test_Case_Merge_08-03-18.xlsx](/uploads/c8a27f0153459850e7a80ec89f4dedfb/Ruby_Test_Case_Merge_08-03-18.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 4, 2018, 12:13

@pchaudhary @smasih @tim.traylor

Please discuss with @tim.traylor why this has not been released to production.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Aug 6, 2018, 05:16

Hi [Kyle] ( @kdjstudios ) & [Tim] (@tim.traylor ),  

We were focused on the Ruby Upgrade on our Saturday release so that if there's any conflict after the ruby upgrade we can resolve that on our window till Monday morning. But fortunately, there was no show stopper & we are good to go.

We have released Validation functionality on both of the Production servers. We would also like to mention that we have also tested some random validations & those are working properly.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 13, 2018, 11:37

@pchaudhary 

I just quickly tested one account and it appears to be working.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 14, 2018, 02:21

thanks for the update Kyle

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 16, 2018, 11:44

closed
